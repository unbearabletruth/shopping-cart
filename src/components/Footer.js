import "../assets/styles/Footer.css"
import instagramIcon from "../assets/images/instagram.svg"
import twitterIcon from "../assets/images/twitter.svg"

const Footer = () => {
    return (
        <div id="footer">
            <ul id="footerInfo">
                <li className="footerInfoList">About us</li>
                <li className="footerInfoList">Polices</li>
                <li className="footerInfoList">Help</li>
                <li className="footerInfoList">Want to work with us?</li>
            </ul>
            <div id="socials">
                <p>Our socials</p>
                <div id="icons">
                    <img src={instagramIcon} alt="instagram" className="socialsIcon"></img>
                    <img src={twitterIcon} alt="twitter" className="socialsIcon"></img>
                </div>
            </div>
        </div>
    )
}

export default Footer;