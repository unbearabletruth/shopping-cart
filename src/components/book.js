import "../assets/styles/Book.css"
import React, { useEffect, useState } from "react";
import data from "../data";
import { useParams } from "react-router-dom";
import { Link } from "react-router-dom"

const Book = ({addToCart}) => {
  const {id} = useParams();
  const [currentBook, setCurrentBook] = useState(data.find((item) => item.id === id))
  
  useEffect(() => {
    const book = data.find((item) => item.title === id);
    setCurrentBook(book);
  }, [id]);

  const handleAmount = (e) => {
    setCurrentBook({
      ...currentBook,
      amount: e.target.value
    })
  }

  const increase = () => {
    setCurrentBook({
      ...currentBook,
      amount: currentBook.amount + 1
    })
  }

  const decrease = () => {
    if (currentBook.amount > 1){
      setCurrentBook({
        ...currentBook,
        amount: currentBook.amount - 1
      })
    }
  }

  return (
    currentBook !== undefined ?
    <div key={currentBook.id} className="bookWrapper">
      <div className="book">
        <img 
          src={currentBook.image} 
          alt={currentBook.title} 
          className="bookImage"
        >
        </img>
        <div className="bookInfo">
          <p className="title">{currentBook.title}</p>
          <p className="author">{currentBook.author}</p>
          <p className="description">{currentBook.description}</p>
          <p className="price">Price: ${currentBook.price.toFixed(2)}</p>
          <div className="inputWrapper">
            <button 
              onClick={() => addToCart(currentBook)} 
              className="toCartButton"
            >
              Add to cart
            </button>
            <label for="amountInput">Amount</label>
            <input 
              id="amountInput"
              className="inputInShop" 
              name={currentBook.id} 
              onChange={handleAmount} 
              value={currentBook.amount} 
              type="number"
            >
            </input>
            <div className="inputButtons">
              <button 
                className="inputButtons" 
                name={currentBook.id} 
                onClick={increase}
              >
                +
              </button>
              <button 
                className="inputButtons" 
                name={currentBook.id} 
                onClick={decrease}
              >
                -
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
     : 
    <BookNotFound/>
  );
};

const BookNotFound = () => {
  return (
      <div id="bookNotFound">
          <p id="notFoundMessage">Sorry, but the book is not found!</p>
          <Link to="/shop" id="backToShop">Look for another book</Link>
      </div>
  )
}

export {Book, BookNotFound};