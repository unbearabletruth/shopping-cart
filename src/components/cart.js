import "../assets/styles/Cart.css"
import { Link } from "react-router-dom"
import deleteIcon from "../assets/images/bin.svg"

const Cart = ({cart, increase, decrease, remove, removeAll, change}) => {
  const totalPrice = (cart.reduce((sum, {amount, price}) => sum + amount * price, 0));
  
  return (
    cart.length !== 0 ?
    <div id="cartWrapper">
      <div id="cart">
      <button id="removeAllButton" onClick={removeAll}>remove all</button>
      {cart.map(book => {
        return(
          <div key={book.id} className="bookInCart" data-testid="bookInCart">
            <img src={book.image} alt={book.title} className="cartBookImage"></img>
            <div className="titleAndAuthor">
              <p className="titleCart">{book.title}</p>
              <p className="authorCart">{book.author}</p>
            </div>
            <div className="priceBlock">
              {book.amount > 1 ?
                <>
                  <p className="priceForOne">One: ${book.price.toFixed(2)}</p>
                  <p>Total: {`$${(book.price * book.amount).toFixed(2)}`}</p>
                </>
              : 
                <p className="priceCart">${book.price.toFixed(2)}</p>}
              
            </div>
            <div className="inputBlock">
              <button className="cartInputButton" name={book.id} onClick={decrease}>-</button>
              <input className="inputInShop" name={book.id} onChange={(e) => change(e, book)} value={book.amount} type="number"></input>
              <button className="cartInputButton" name={book.id} onClick={increase}>+</button>
            </div>
            <img src={deleteIcon} alt="deleteIcon" className="deleteImage" name={book.id} onClick={remove} aria-label={book.author}></img> 
          </div>
        )
        })}
        </div>
          <div id="payment">
            <p id="totalPriceText">Order total</p>
            <p id="totalPrice">${totalPrice.toFixed(2)}</p>
            <button id="payButton">Pay</button>
          </div>
    </div>
    :
    <div id="emptyCartInfo">
      <p id="emptyCartText">Your cart is empty</p>
      <Link to="/shop" id="cartToShop">Fill it in!</Link>
    </div>
        
  );
};
  
export default Cart;