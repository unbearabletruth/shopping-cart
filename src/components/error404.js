import { Link } from "react-router-dom"
import "../assets/styles/error404.css"

const Error404 = () => {
    return(
        <div id="errorWrapper">
            <p id="errorNumber">404</p>
            <p id="errorMessage">Sorry, but your page is not found</p>
            <Link to="/" id="errorToHome">Home</Link>
        </div>
    )
}

export default Error404;