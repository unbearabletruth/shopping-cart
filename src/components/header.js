import { Link, useNavigate } from "react-router-dom"
import "../assets/styles/Header.css"
import cartIcon from "../assets/images/cart.png"
import searchIcon from "../assets/images/magnifyingGlass.svg"
import { useState } from "react"
import data from "../data";

const Header = ({cart}) => {
  const [search, setSearch] = useState("");
  const navigate = useNavigate();
  
  const handleSearch = (e) => {
    setSearch(e.target.value)
  }

  const onSearchSubmit = () => {
    if (data.some(item => item.title === search)){
      navigate(`./shop/${search}`)
    }else{
      navigate('./shop/*')
    }
  }

  const onSearchSubmitEnter = (e) => {
    if (e.key === "Enter" && data.some(item => item.title === search)){
      navigate(`./shop/${search}`)
    }
    else if (e.key === "Enter"){
      navigate('./shop/*')
    }
  }

    return(
        <nav>
          <ul id="headerLinks">
            <li><Link to="/">Home</Link></li>
            <li><Link to="/shop">Shop</Link></li>
          </ul>
          <div id="searchAndButton">
            <input 
              id="headerSearch" 
              onChange={handleSearch} 
              onKeyDown={onSearchSubmitEnter} 
              type="search"
            >
            </input>
            <button id="searchButton" onClick={onSearchSubmit}>
              <img id="searchIcon" src={searchIcon} alt="search"></img>
            </button>
          </div>
          <Link to="/cart" id="cartLink">
            <img id="cartIcon" src={cartIcon} alt="cart icon"></img>
            <div id="cartItemsNumber">{cart.length}</div>
          </Link>
        </nav>
    )
}

export default Header;