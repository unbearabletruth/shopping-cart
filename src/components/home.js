import background from "../assets/images/homeBackground.jpg"
import "../assets/styles/Home.css"
import { Link } from "react-router-dom"

const Home = () => {
    return (
      <>
        <img id="homeBackground" src={background} alt="background"></img>
        <div id="homeInfo">
          <p id="homeText">Check out our books!</p>
          <Link to="/shop" id="homeToShop">Shop</Link>
        </div>
      </>
    );
  };
  
export default Home;