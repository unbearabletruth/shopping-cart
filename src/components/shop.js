import "../assets/styles/shop.css"
import React, { useState } from "react";
import data from "../data";

const Shop = ({addToCart}) => {
 

    return (
      <>
        {data.map(book => {
          return (
            <div key={book.id} id="bookShop">
              <div className="bookCard">
                <Book image={book.image} title={book.title} author={book.author}/>
                <button onClick={() => addToCart(book)}>Add to cart</button>
              </div>
            </div>
          )
        })} 
      </>
    );
  };

const Book = ({image, title, author}) => {
  return (
    <>
      <div className="bookInfo">
          <img src={image} alt={title} className="shopBookImage"></img>
          <p>{title}</p>
          <p>{author}</p>
          <input type="number"></input>
      </div>
    </>
  );
}

export default Shop;