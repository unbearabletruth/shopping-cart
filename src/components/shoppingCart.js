const ShoppingCart = () => {
    return (
      <>
        <div>
            <h1>Hello from shoppingCart</h1>
        </div>
        
      </>
    );
  };
  
export default ShoppingCart;