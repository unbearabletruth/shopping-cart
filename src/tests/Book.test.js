import { render, screen } from '@testing-library/react';
import {Book} from '../components/Book';
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from 'react-router-dom';
import Router from "react-router-dom";

jest.mock('../data', () => ([
    { image: 'smth', title: "Behave", author: "author", description: "description", amount: 1, price: 10, id: 1 },
  ]))

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
 }));

const addToCart = jest.fn();

describe('book', () => {
  describe('renders proper DOM elements', () => {
    test('snapshot of page with book', () => {
      jest.spyOn(Router, 'useParams').mockReturnValue({ id: 'Behave' })
      const {container} = render(
        <MemoryRouter>
          <Book />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });

    test('snapshot of book not found', () => {
      jest.spyOn(Router, 'useParams').mockReturnValue({ id: '' })
      const {container} = render(
        <MemoryRouter>
          <Book />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });

    test('renders book not found message', () => {
      jest.spyOn(Router, 'useParams').mockReturnValue({ id: '' })
      render(
        <MemoryRouter>
            <Book />
        </MemoryRouter>
        )
      expect(screen.getByText('Sorry, but the book is not found!')).toBeInTheDocument();
    });

    test('renders book', () => {
      jest.spyOn(Router, 'useParams').mockReturnValue({ id: 'Behave' })
      render(
        <MemoryRouter>
            <Book />
        </MemoryRouter>
        )
      expect(screen.getByRole("img", {name: "Behave"})).toBeInTheDocument();
    });
  });

  describe('component functionality', () => {
    test('increments/decrements amount to add', async () => {
      jest.spyOn(Router, 'useParams').mockReturnValue({ id: 'Behave' })
      const user = userEvent.setup();
      render(
        <MemoryRouter>
            <Book />
        </MemoryRouter>
        )
      const buttonInc = screen.getByRole('button', {name: "+"});
      const buttonDec = screen.getByRole('button', {name: "-"});
      const amountInput = screen.getByLabelText("Amount");
      await user.click(buttonInc);
      expect(amountInput).toHaveValue(2);
      await user.click(buttonDec);
      expect(amountInput).toHaveValue(1);
    });

    test('sets amount to add', async () => {
      jest.spyOn(Router, 'useParams').mockReturnValue({ id: 'Behave' })
      const user = userEvent.setup();
      render(
        <MemoryRouter>
          <Book/>
        </MemoryRouter>
      )
      const amountInput = screen.getByLabelText("Amount");
      await user.type(amountInput, "{backspace}5");
      expect(amountInput).toHaveValue(5);
    })

    test('adds to cart on button click', async () => {
      jest.spyOn(Router, 'useParams').mockReturnValue({ id: 'Behave' })
      const user = userEvent.setup();
      render(
        <MemoryRouter>
          <Book
            addToCart={addToCart}
          />
        </MemoryRouter>
      )
      const button = screen.getByRole('button', {name: "Add to cart"});
      await user.click(button);
      expect(addToCart).toHaveBeenCalledTimes(1);
    })
  });
});