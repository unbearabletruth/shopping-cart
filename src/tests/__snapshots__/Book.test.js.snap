// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book renders proper DOM elements snapshot of book not found 1`] = `
<div>
  <div
    id="bookNotFound"
  >
    <p
      id="notFoundMessage"
    >
      Sorry, but the book is not found!
    </p>
    <a
      href="/shop"
      id="backToShop"
    >
      Look for another book
    </a>
  </div>
</div>
`;

exports[`book renders proper DOM elements snapshot of page with book 1`] = `
<div>
  <div
    class="bookWrapper"
  >
    <div
      class="book"
    >
      <img
        alt="Behave"
        class="bookImage"
        src="smth"
      />
      <div
        class="bookInfo"
      >
        <p
          class="title"
        >
          Behave
        </p>
        <p
          class="author"
        >
          author
        </p>
        <p
          class="description"
        >
          description
        </p>
        <p
          class="price"
        >
          Price: $
          10.00
        </p>
        <div
          class="inputWrapper"
        >
          <button
            class="toCartButton"
          >
            Add to cart
          </button>
          <label
            for="amountInput"
          >
            Amount
          </label>
          <input
            class="inputInShop"
            id="amountInput"
            name="1"
            type="number"
            value="1"
          />
          <div
            class="inputButtons"
          >
            <button
              class="inputButtons"
              name="1"
            >
              +
            </button>
            <button
              class="inputButtons"
              name="1"
            >
              -
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
